File: network-topology-mapper/backend/app/routers/topology.py
from fastapi import APIRouter, Query
from typing import Optional

from app.services.graph.graph_builder import graph_builder
from app.db.topology_db import topology_db
from app.services.graph.resilience_scorer import resilience_scorer

router = APIRouter(prefix="/api", tags=["topology"])


@router.get("/topology")
def get_topology(
    layer: Optional[str] = Query(None, description="physical|logical|application"),
    vlan: Optional[int] = Query(None),
    subnet: Optional[str] = Query(None),
    device_type: Optional[str] = Query(None),
    risk_min: Optional[float] = Query(None),
):
    topology = graph_builder.get_full_topology(
        layer=layer, vlan=vlan, subnet=subnet,
        device_type=device_type, risk_min=risk_min,
    )
    return topology


@router.get("/topology/stats")
def get_topology_stats():
    topology = graph_builder.get_full_topology()
    devices = topology.get("devices", [])
    connections = topology.get("connections", [])

    online = sum(1 for d in devices if d.get("status") == "online")
    offline = sum(1 for d in devices if d.get("status") == "offline")
    degraded = sum(1 for d in devices if d.get("status") == "degraded")

    active_conns = sum(1 for c in connections if c.get("status", "active") == "active")

    type_counts = {}
    for d in devices:
        dt = d.get("device_type", "unknown")
        type_counts[dt] = type_counts.get(dt, 0) + 1

    avg_risk = sum(d.get("risk_score", 0) for d in devices) / max(len(devices), 1)

    # Get resilience info
    try:
        resilience = resilience_scorer.calculate_global_resilience()
        risk_score = resilience.get("score", 0)
        spof_count = resilience.get("spof_count", 0)
    except Exception:
        risk_score = round(avg_risk * 10, 1)
        spof_count = 0

    return {
        "total_devices": len(devices),
        "total_connections": len(connections),
        "online": online,
        "offline": offline,
        "degraded": degraded,
        "active_connections": active_conns,
        "connection_uptime_pct": round(active_conns / max(len(connections), 1) * 100, 1),
        "type_counts": type_counts,
        "risk_score": risk_score,
        "spof_count": spof_count,
        "avg_device_risk": round(avg_risk, 3),
    }


@router.post("/topology/clear")
def clear_topology():
    topology_db.clear_all()
    return {"status": "ok", "message": "All devices, connections, and dependencies cleared."}


@router.get("/devices/{device_id}")
def get_device(device_id: str):
    device = topology_db.get_device(device_id)
    if not device:
        # Search in topology
        topology = graph_builder.get_full_topology()
        for d in topology.get("devices", []):
            if d.get("id") == device_id:
                device = d
                break
    if not device:
        return {"error": "Device not found"}

    # Get risk details
    risk_details = resilience_scorer.calculate_device_risk(device_id)
    device["risk_details"] = risk_details

    return device


@router.get("/devices/{device_id}/connections")
def get_device_connections(device_id: str):
    topology = graph_builder.get_full_topology()
    connections = [
        c for c in topology.get("connections", [])
        if c.get("source_id") == device_id or c.get("target_id") == device_id
    ]
    return {"device_id": device_id, "connections": connections}


@router.get("/devices/{device_id}/dependencies")
def get_device_dependencies(device_id: str):
    topology = graph_builder.get_full_topology()
    deps = topology.get("dependencies", [])

    upstream = [d for d in deps if d.get("source_id") == device_id]
    downstream = [d for d in deps if d.get("target_id") == device_id]

    return {
        "device_id": device_id,
        "depends_on": upstream,
        "depended_by": downstream,
    }

File: network-topology-mapper/frontend/src/types/topology.ts
export type DeviceType =
  | 'router'
  | 'switch'
  | 'server'
  | 'workstation'
  | 'firewall'
  | 'ap'
  | 'printer'
  | 'iot'
  | 'unknown';

export type DeviceStatus = 'online' | 'offline' | 'degraded';

export type ConnectionType = 'ethernet' | 'fiber' | 'wireless' | 'vpn' | 'virtual';

export type ConnectionStatus = 'active' | 'disabled' | 'degraded' | 'flapping';

export type AlertSeverity = 'critical' | 'high' | 'medium' | 'low' | 'info';

export type AlertType =
  | 'new_device'
  | 'topology_change'
  | 'spof'
  | 'anomaly'
  | 'flapping'
  | 'device_offline'
  | 'config_drift';

export type AlertStatus = 'open' | 'acknowledged' | 'resolved' | 'dismissed';

export type ScanType = 'active' | 'passive' | 'snmp' | 'full';

export type ScanStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

export type LayerType = 'physical' | 'logical' | 'application';

export type LayoutType = 'cola' | 'dagre' | 'circle' | 'grid';

export interface Device {
  id: string;
  ip: string;
  mac: string;
  hostname: string;
  device_type: DeviceType;
  vendor: string;
  model: string;
  os: string;
  open_ports: number[];
  services: string[];
  first_seen: string;
  last_seen: string;
  discovery_method: string;
  vlan_ids: number[];
  subnet: string;
  location: string;
  risk_score: number;
  criticality: 'high' | 'medium' | 'low';
  is_gateway: boolean;
  status: DeviceStatus;
  tier: number;
  risk_details?: RiskDetails;
}

export interface RiskDetails {
  device_id: string;
  risk_score: number;
  factors: string[];
}

export interface Connection {
  id: string;
  source_id: string;
  target_id: string;
  connection_type: ConnectionType;
  bandwidth: string;
  switch_port: string;
  vlan: number | null;
  latency_ms: number;
  packet_loss_pct: number;
  is_redundant: boolean;
  protocol: string;
  status: ConnectionStatus;
  first_seen: string;
  last_seen: string;
}

export interface Dependency {
  id: string;
  source_id: string;
  target_id: string;
  dependency_type: string;
  service_port: number;
  criticality: string;
  discovered_via: string;
}

export interface Topology {
  devices: Device[];
  connections: Connection[];
  dependencies: Dependency[];
}

export interface TopologyStats {
  total_devices: number;
  total_connections: number;
  online: number;
  offline: number;
  degraded: number;
  active_connections: number;
  connection_uptime_pct: number;
  type_counts: Record<string, number>;
  risk_score: number;
  spof_count: number;
  avg_device_risk: number;
}

export interface Alert {
  id: string;
  alert_type: AlertType;
  severity: AlertSeverity;
  title: string;
  description: string;
  device_id: string | null;
  details: Record<string, any>;
  created_at: string;
  acknowledged_at: string | null;
  resolved_at: string | null;
  status: AlertStatus;
}

export interface Scan {
  id: string;
  scan_type: ScanType;
  status: ScanStatus;
  started_at: string | null;
  completed_at: string | null;
  target_range: string;
  devices_found: number;
  new_devices: number;
  config: Record<string, any>;
  progress?: ScanProgress;
}

export interface ScanProgress {
  scan_id: string;
  percent: number;
  phase: string;
  devices_found: number;
  log_messages?: string[];
}

export interface SPOF {
  device_id: string;
  hostname: string;
  device_type: string;
  impact: number;
  reason: string;
  risk_score: number;
}

export interface SimulationResult {
  blast_radius: number;
  blast_radius_pct: number;
  disconnected_devices: SimulatedDevice[];
  degraded_devices: SimulatedDevice[];
  safe_device_ids: string[];
  removed_node_ids: string[];
  affected_services: string[];
  risk_delta: number;
  new_components: number;
  recommendations: Recommendation[];
}

export interface SimulatedDevice {
  id: string;
  hostname: string;
  device_type: string;
  status: string;
  lost_connections?: number;
}

export interface Recommendation {
  action: string;
  impact: string;
  priority: string;
}

export interface ResilienceReport {
  generated: boolean;
  score: number;
  max_score: number;
  report: string;
  spofs: SPOF[];
  stats: { total_devices: number; total_connections: number };
}

export interface ResilienceScore {
  score: number;
  max_score: number;
  breakdown: {
    connectivity: number;
    redundancy: number;
    spof_resistance: number;
    path_diversity: number;
  };
  total_devices: number;
  total_connections: number;
  spof_count: number;
  spofs: SPOF[];
}

export interface TopologySnapshot {
  id: string;
  created_at: string;
  device_count: number;
  connection_count: number;
  risk_score: number;
  snapshot_data: Record<string, any>;
}

export interface WSMessage {
  type: string;
  data: any;
}

export type ViewType =
  | 'topology'
  | 'dashboard'
  | 'scan'
  | 'simulate'
  | 'alerts'
  | 'heatmap'
  | 'timeline'
  | 'reports'
  | 'settings';
